packages/fluentui/docs/src/prototypes/VirtualizedTree/VirtualTree.tsx

```diff
@@ -0,0 +1,178 @@
+}
+
+export const VirtualTree: ComponentWithAs<'div', VirtualTreeProps> = props => {
+  const context = useFluentContext();
+  const { setStart, setEnd } = useTelemetry(VirtualTree.displayName, context.telemetry);
+  setStart();
+
+  const { children, className, design, styles, variables, height, estimatedItemSize } = props;
+
+  const ElementType = getElementType(props);
+
+  const unhandledProps = useUnhandledProps([...Tree.handledProps, 'estimatedItemSize', 'itemSize'], props);
+
+  const getA11yProps = useAccessibility(props.accessibility, {
+    debugName: VirtualTree.displayName,
+    rtl: context.rtl,
+  });
+
+  const { classes } = useStyles<TreeStylesProps>(Tree.displayName, {
+    className: treeClassName,
+    mapPropsToInlineStyles: () => ({
+      className,
+      design,
+      styles,
+      variables,
+    }),
+    rtl: context.rtl,
+  });
+
+  const {
+    visibleItemIds,
+    getItemById,
+    registerItemRef,
+    toggleItemActive,
+    focusItemById,
+    expandSiblings,
+    listRef,
+  } = useVirtualTree(props);
+
+  const contextValue: TreeRenderContextValue = React.useMemo(
+    () => ({
+      getItemById,
+      registerItemRef,
+      toggleItemActive,
+      focusItemById,
+      expandSiblings,
+      toggleItemSelect: _.noop,
+    }),
+    [getItemById, registerItemRef, toggleItemActive, focusItemById, expandSiblings],
+  );
+
+  const getItemKey = React.useCallback((index: number, data: VirtualItemData) => data.visibleItemIds[index], []);
+
+  const getItemSize = (index: number) => {
+    const id = visibleItemIds[index];
+    return getItemById(id).item.itemSize || estimatedItemSize;
+  };
+  React.useLayoutEffect(() => {
+    (listRef.current as any)?.resetAfterIndex(0);
+  }, [listRef, visibleItemIds]); // when item collapsed/expanded (visibleItemIds change), refresh react-window itemSize cache
+
+  const createTreeItem = React.useCallback(
+    (id, style) => {
+      const item = getItemById(id);
+      if (item) {
+        const { expanded, parent, level, index, treeSize } = item;
+        return TreeItem.create(item.item, {
+          defaultProps: () =>
+            getA11yProps('item', {
+              renderItemTitle: props.renderItemTitle,
+            }),
+          overrideProps: {
+            style, // came from react-window
+            expanded,
+            parent,
+            key: id,
+            level,
+            index,
+            treeSize,
+            selectable: false,
+          },
+        });
+      }
+      return null;
+    },
+    [getA11yProps, getItemById, props.renderItemTitle],
+  );
+
+  const innerElementType = React.useMemo(
+    () => React.forwardRef((props, ref) => <div ref={ref as React.RefObject<HTMLDivElement>} {...props} role="none" />),
+    [],
+  );
+  const outerElementType = React.useMemo(
+    () => React.forwardRef((props, ref) => <div ref={ref as React.RefObject<HTMLDivElement>} {...props} role="none" />),
+    [],
+  );
```

Comment:
Let's extract these two to real components, it also can be typed without type casts

```tsx
const OuterElement = React.forwardRef<HTMLDivElement>((props, ref) => <div ref={ref} {...props} role="none" />);
```